lib/workload/stateful/filemanager/database/queries/ingester/aws/update_reordered_for_created.sql

```diff
@@ -0,0 +1,118 @@
+-- Update the matching s3_objects which should be re-ordered based on the created event. Returns the
+-- data associated with the event before the update, if an update occurred.
+
+-- First, unnest the input parameters into a query.
+with input as (
+    select
+        *
+    from unnest(
+        $1::uuid[],
+        $2::text[],
+        $3::text[],
+        $4::timestamptz[],
+        $5::integer[],
+        $6::text[],
+        $7::timestamptz[],
+        $8::text[],
+        $9::storage_class[],
+        $10::text[],
+        $11::text[]
+    ) as input (
+        s3_object_id,
+        bucket,
+        key,
+        created_date,
+        size,
+        checksum,
+        last_modified_date,
+        e_tag,
+        storage_class,
+        version_id,
+        created_sequencer
+    )
+),
+-- Then, select the objects that need to be updated.
+current_objects as (
+    select
+        s3_object.*,
+        input.s3_object_id as input_id,
+        input.bucket as input_bucket,
+        input.key as input_key,
+        input.version_id as input_version_id,
+        input.created_sequencer as input_created_sequencer,
+        input.created_date as input_created_date,
+        input.size as input_size,
+        input.checksum as input_checksum,
+        input.last_modified_date as input_last_modified_date,
+        input.e_tag as input_e_tag,
+        input.storage_class as input_storage_class
+    from s3_object
+    -- Grab the relevant values to update with.
+    join input on
+        input.bucket = s3_object.bucket and
+        input.key = s3_object.key and
+        input.version_id = s3_object.version_id
+    -- Lock this pre-emptively for the update.
+    for update
+),
+-- And filter them to the objects that need to be updated.
+objects_to_update as (
+    select
+        *
+    from current_objects
+    where
+        -- Check the sequencer condition. We only update if there is a created
+        -- sequencer that is closer to the deleted sequencer.
+        current_objects.deleted_sequencer > current_objects.input_created_sequencer and
+        (
+            -- Updating a null sequencer doesn't cause the event to be reprocessed.
+            current_objects.created_sequencer is null or
+            -- If a sequencer already exists this event should be reprocessed because this
+            -- sequencer could belong to another object.
+            current_objects.created_sequencer < current_objects.input_created_sequencer
+        )
+        -- And there should not be any objects with a created sequencer that is the same as the input created
+        -- sequencer because this is a duplicate event that would cause a constraint error in the update.
+        and current_objects.input_created_sequencer not in (
+            select created_sequencer from current_objects where created_sequencer is not null
+        )
+),
```

Comment:
Given that this is a bit of a monster statement. I think the understandability would be greatly improved if this was split across multiple SQL queries, where the application code did the majority of this sequencer logic/comparison. However, this would be at the cost of more application <-> database calls, although I'm not sure if this matters that much for now.

Reply:
Doesn't look that bad? The comments make it more understandable... how large would the perf impact be if you split?

Reply:
I can't image this cost would be that great. It's not like it's an order of magnitude thing, I think it would be down to how well the database could optimize the bigger query.

Reply:
Up to you if you want to refactor, I'd leave it as-is... it's more readable today than it was yesterday (to me) so this piece of code complexity is a bit subjective in this case, I just needed to be in the right frame of mind to understand it and its context :-S

That being said, I think that an ARCHITECTURE.md explanation would be helpful regardless: https://matklad.github.io/2021/02/06/ARCHITECTURE.md.html